### ISSUE TITLE
Question about  perturbing cortex across individuals.

### ISSUE THREAD

**Issue by zhibotian** [jingyanghuo/neuropictor#2]
Hi Jingyang,

Congratulations on having your paper accepted at ECCV! It’s truly an interesting piece of work.

I am particularly intrigued by the experiments in **Section 2.5: Perturbing Cortex Across Individuals**. Could you please explain how you replaced the fMRI signals from specific regions of subject-1 with those from subject-2? Specifically, are the V1, V2, and other ROIs in the voxels of the same length across subjects?

Looking forward to your response and thank you in advance!

Best regards,
Zhibo

**Comment by zhibotian** [comment]
Hi Jingyang,

Sorry for the confusion in my previous quesion. I meant to refer to Section 2.5 in the Supplementary Material.

Thanks again for your time and I look forward to your insights!

Best regards,
Zhibo

**Comment by jingyanghuo** [comment]
Hi Zhibo,

Thank you for your interest in our work.

We have a predefined cortical regions surface segmentation (`vc_masks.npy`). You can download it from [this link](https://huggingface.co/Fudan-fMRI/neuropictor/blob/main/NSD/vc_masks.npy). For the fMRI surfaces of two subjects viewing the same image, we exchanged specific regions between them (e.g., swapping the V1 region of sub-1 with that of sub-2). The regions are divided according to the "Cortical Division" in Table 1 of [this paper](https://link.springer.com/article/10.1007/s00429-021-02421-6). Here is the breakdown:


- **Primary Visual**: ['V1']
- **Early Visual**: ['V2', 'V3', 'V4']
- **Dorsal Stream Visual**: ['V3A', 'V3B', 'V6', 'V6A', 'V7']
- **Ventral Stream Visual**: ['FFC', 'PIT', 'V8', 'VMV1', 'VMV2', 'VMV3', 'VVC']
- **MT Complex**: ['LO1', 'LO2', 'LO3', 'FST', 'MT', 'MST', 'PH', 'V3CD', 'V4t']
- **Medial Temporal**: ['PHA1', 'PHA2', 'PHA3']
- **Inferior Parietal**: ['IP0']

Each region corresponds to a specific number of surface pixels:
- Primary Visual: 7060 pixels
- Early Visual: 8742 pixels
- Dorsal Stream Visual: 2550 pixels
- Ventral Stream Visual: 3806 pixels
- MT Complex: 3784 pixels
- Medial Temporal: 1273 pixels
- Inferior Parietal: 435 pixels

In our paper, we grouped these regions into three functional categories:
1. **Primary Visual**
2. **Early Visual**
3. **Some Higher Visual** (which combines Dorsal Stream Visual, Ventral Stream Visual, MT Complex, Medial Temporal, and Inferior Parietal regions)

Below is a code example for exchanging specific regions between two fMRI surfaces:

```python
Primary_Visual = ['V1']

Early_Visual = ['V2','V3','V4']

Dorsal_Stream_Visual = ['V3A','V3B','V6','V6A','V7']

Ventral_Stream_Visual = ['FFC','PIT','V8','VMV1','VMV2','VMV3','VVC']

MT_Complex = ['LO1','LO2','LO3','FST','MT','MST','PH','V3CD','V4t']

Medial_Temporal = ['PHA1','PHA2','PHA3']

Inferior_Parietal = ['IP0']

"""
Inputs: 
exchange_roi_names = Primary_Visual
# exchange_roi_names = Early_Visual
# exchange_roi_names =  Dorsal_Stream_Visual+Ventral_Stream_Visual+MT_Complex+Medial_Temporal+Inferior_Parietal
surface = np.load() # fmri surface from sub-1
exchange_surface = np.load() # fmri surface from sub-2
"""

# Function to create a mask based on a list of ROI names.
def mask_with_vc_roi(roi_names):
    mask = np.zeros_like(self.vc_roi_mask, dtype=bool)
    for roi_name in roi_names:
        roi_level_value = roi_dict[roi_name]
        roi_mask = vc_roi_mask == roi_level_value
        mask = mask | roi_mask
    return mask

# List of region of interest (ROI) names in the brain.
roi_names = ['V1', 'V2', 'V3', 'V3A', 'V3B', 'V3CD', 'V4', 'LO1', 'LO2', 'LO3', 
             'PIT', 'V4t', 'V6', 'V6A', 'V7', 'V8', 'PH', 'FFC', 'IP0', 'MT', 
             'MST', 'FST', 'VVC', 'VMV1', 'VMV2', 'VMV3', 'PHA1', 'PHA2', 'PHA3']

# Dictionary to store region information.
roi_dict = {name: idx for idx, name in enumerate(roi_names, 1)}

# Load the visual cortex (VC) mask from a .npy file.
vc_roi_mask = np.load("path/to/vc_masks.npy")

# Generate a mask for the specified ROIs.
exchange_mask = mask_with_vc_roi(exchange_roi_names)

# Use the exchange mask to update the surface.  If exchange_mask is True, take values from exchange_surface; otherwise, from surface.
new_surface = np.where(exchange_mask, exchange_surface, surface)
```

I hope this helps. Please feel free to reach out if you have more questions.

Best regards,  
Jingyang

**Comment by zhibotian** [comment]
Hi Jingyang,

Thank you very much for your detailed explanation! It has fully clarified my questions.

Best regards,
zhibo

**Comment by zhibotian** [comment]
Hi Jingyang,

Thank you again for your previous explanation. I have another question regarding the preprocessing method. In the previous approach, the fMRI data is reduced to a one-dimensional representation (e.g., 15,724 for subject-1). Can the surface pixels of each region be directly mapped to this one-dimensional fMRI data? As you mentioned, each region has a specific number of surface pixels, but when summed across all regions, does this total match the full fMRI data? It seems like the total doesn't equal 256*256. Could you please help me understand this?

Thanks a lot for your help!

Best regards,
Zhibo

**Comment by jingyanghuo** [comment]
Thank you for your continued interest and thoughtful questions.

1. **Can the surface pixels of each region be directly mapped to this one-dimensional fMRI data?**  
   This is difficult, as mapping from group-level surface representations back to individual-level space is challenging. The 15,724 fmri for subject-1 was derived as a subset of voxels in the posterior cortex most responsive to the visual stimuli presented, not based on predefined cortical regions. Therefore, the 1D fMRI data cannot be directly matched across individuals.

2. **It seems like the total doesn't equal 256*256.**  
   You can refer to Fig. 7 in our paper. We padded the surface of the brain regions to 256x256 by filling the non-VC (visual cortex) areas with zeros, which is why the full fMRI data does not equal the total surface pixel count.

3. Additionally, if you are interested in 1D representations, we also have the full-brain 1D voxels in the fsLR group space (about 90,000+ voxels), from which you can extract specific regions. We also have the 1D VC voxels pre-extracted if you need them. Note that these 1D voxels are all in group space, and they are not the same as those in individual space. I can upload this data for your research if needed.

**Comment by zhibotian** [comment]
Thank you for your clear response. I also read your work on fMRI-PTE, where the fMRI data seems to be mapped onto 1023x2514, and you processed it into 256x256. Would this transformation impact the results?

**Comment by jingyanghuo** [comment]
The 1023x2514 map represents the full-brain fMRI data. In processing the data, we first select a subset using predefined VC regions, which gives us a map with dimensions of 693x714. We then apply padding and resizing to transform it into 256x256. This compression was done to manage computational costs.

I can't say for certain whether this transformation impacts the results, as we haven't conducted experiments specifically to test that. I guess that a higher resolution might slightly improve the results. If you want to explore this further, you could fine-tune the encoder using the original 693x714 data on NSD for quicker validation, since the training costs with UKB are quite high. If you’d like to investigate this, we can provide additional data and code.

**Comment by zhibotian** [comment]
Hello, I have a question about why it's necessary to load the pre-trained encoder (apparently trained on the UKB dataset) before running the demo. Is the encoder kept frozen during training on multiple and single subjects, or is it updated during those stages as well?

**Comment by jingyanghuo** [comment]
The fMRI encoder is updated during training in the fMRI-to-image process (both for multiple and single subjects) and is not frozen. You can run the demo without loading the pre-trained UKB encoder. For inference, you can comment out lines 56-66 in cldm/cldm_fmri_surf.py where the FMRI Encoder weights are loaded. The demo code requires loading the FMRI Encoder weights because I haven't made a distinction for this in the earlier code. This is a minor oversight; sorry for the confusion.

**Comment by zhibotian** [comment]
Thank you for your reply. So, if I load the entire model checkpoint (e.g., epoch_015.pth), does it include the encoder’s parameters that have been trained with NSD?

**Comment by jingyanghuo** [comment]
Yes, the encoder’s parameters is included in the checkpoint.

**Comment by jingyanghuo** [comment]
Thanks. Same to you.

**Comment by zhibotian** [comment]
Hello, I hope you're doing well.  I have a question regarding your work. In your approach, you transform fMRI data into unified 2D brain activation images. I was wondering if you have ever considered using a 1D voxel method instead? Would this approach potentially lead to a significant decrease in the results? Thanks.

**Comment by jingyanghuo** [comment]
This project is initially based on using a 2D surface and a pretrained autoencoder in UKB, so we don’t conduct a complete experiment with 1D voxel data. I can’t provide quantitative results for that. However, in some earlier small experiments, like directly using regression models such as brain-diffuser, we find that the difference between 1D voxel and 2D surface results is minimal. I believe that the information in 1D voxel and 2D surface is quite similar, as the 2D surface is derived from 1D voxel through deformation interpolation.

For this project, if we use 1D voxel data, we need: (1) A unified pretrained autoencoder for 1D voxel (you can refer to MBM in MindVis for this); (2) Alignment of 1D voxel data across subjects for multi-subject pretraining. These tasks could be time-consuming. With these conditions and a similar number of parameters, I think it would achieve similar/comparable results.
